### PR TITLE
test(edge-cases): generated inputs and real concurrent processes

### PR DESCRIPTION
**Pull request by ooples**
The two remaining blind spots behind *"we keep finding defects"*. Both were built by first asking what the tests **could not see**, then checking the new tests can see it.

## 1. Generated inputs (fast-check)

Hand-picked inputs come from the same assumptions that produced the bug. Two recent defects were exactly that shape:

- `'\bnpx\s+jest\b'` in single quotes resolves to a **backspace** followed by `npxs+jest` — the assertion went green against a pattern nobody wrote
- `not.toContain('..')` on a filename, when the real invariant was containment and `a.._..b` satisfies one while failing the other

In both, the author chose the input and the input agreed with him.

**17 properties** over `canonicalPath`, `canonicalKey`, `nodeId`, `contentHash`, symbol/import extraction, payload normalisation and the graph parser — idempotence, determinism, never-throws, spans inside the file, imports always relative, and a graph that survives arbitrary bytes rather than losing everything after one bad line. The nasty-input corpus is drawn from real failures here: the backspace that mangled a trigger, NUL, BOM, UNC and drive-relative paths, mixed separators.

### It immediately found one — in my own test

The containment property failed in 58 runs with the counterexample `".."`. That's **correct behaviour**: `join(root, '..')` resolves to the parent and is supposed to. Containment isn't a property of `join` at all — it's a property of the **sanitiser**, which is what actually defends the boundary. The property now states that, plus a second test pinning that raw `join` really does escape, so nobody later removes the sanitiser believing `join` was safe all along.

## 2. Real concurrent processes

Every test here ran in one process, and **a lock that is never contended always looks correct**. That blind spot cost twice already: a state field dropped on write, and a lock retried 20× with no delay.

### And the honest part

The first version of this suite **passed with the append lock removed** — it proved nothing. Enlarging records to 96 KB didn't change that, so I checked directly:

> Unlocked. 8 processes × 40 records × 256 KB — far past the PIPE_BUF threshold the lock's own comment cites. **320 of 320 lines, none torn.**

Append-mode writes are atomic on Windows, so no run *here* can demonstrate the lock's value.

The suite is kept and its docstring says exactly that. It verifies the invariants under genuine multi-process load; CI runs Linux where PIPE_BUF is real and the same assertions may have teeth. The alternative — a green concurrency test that would stay green with the lock deleted — is evidence of nothing.

**111 suites, 1512 tests.** `fast-check` added as a dev dependency.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/hooks/concurrency.test.mjs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    // A crashed writer must not stop every future write for the life of the
+    // session. The lock is broken once it is provably stale.
+    writeFileSync(join(dir, '.graph.lock'), '');
+    const old = new Date(Date.now() - 60_000);
+    try {
+      const { utimesSync } = require('fs');
+      utimesSync(join(dir, '.graph.lock'), old, old);
+    } catch {
+      /* handled below by the assertion itself */
+    }
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/hooks/concurrency.test.mjs]: Correct, and it was worse than described.

`require` is undefined in a `.mjs` file Jest runs as native ESM, so `require('fs')` threw a `ReferenceError` straight into the catch on every run — the lock was never backdated. Fixed: `utimesSync` is imported at the top, called directly with no catch, and the mtime is asserted afterwards, so a backdate that silently fails can no longer be mistaken for recovery.

Fixing that alone did not make the test meaningful, though. `withLock` calls `write()` whether or not the lock was taken — deliberately, so a crashed writer cannot stop the graph being written — which means "the write still lands" is true with the stale-lock branch deleted entirely. I checked by deleting it, and the test stayed green.

The observable difference is reclamation, not success: a broken stale lock is held and released, so the file is gone afterwards; an unbroken one survives and leaves every later write in the session unsynchronised. That is what the test now asserts, and deleting the branch turns it red. Renamed to `is broken and reclaimed, not merely stepped around`.

Fixed in 8b0ddc4.
